qa/tasks/kafka_failover.py

```diff
@@ -0,0 +1,236 @@
+"""
+Deploy and configure Kafka for Teuthology
+"""
+import contextlib
+import logging
+import time
+import os
+
+from teuthology import misc as teuthology
+from teuthology import contextutil
+from teuthology.orchestra import run
+
+log = logging.getLogger(__name__)
+
+def get_kafka_version(config):
+    for client, client_config in config.items():
+        if 'kafka_version' in client_config:
+            kafka_version = client_config.get('kafka_version')
+    return kafka_version
+
+kafka_prefix = 'kafka_2.13-'
+
+def get_kafka_dir(ctx, config):
+    kafka_version = get_kafka_version(config)
+    current_version = kafka_prefix + kafka_version
+    return '{tdir}/{ver}'.format(tdir=teuthology.get_testdir(ctx),ver=current_version)
+
+
+@contextlib.contextmanager
+def install_kafka(ctx, config):
+    """
+    Downloading the kafka tar file.
+    """
+    assert isinstance(config, dict)
+    log.info('Installing Kafka...')
+
+    # programmatically find a nearby mirror so as not to hammer archive.apache.org
+    apache_mirror_cmd="curl 'https://www.apache.org/dyn/closer.cgi' 2>/dev/null | " \
+        "grep -o '<strong>[^<]*</strong>' | sed 's/<[^>]*>//g' | head -n 1"
+    log.info("determining apache mirror by running: " + apache_mirror_cmd)
+    apache_mirror_url_front = os.popen(apache_mirror_cmd).read().rstrip() # note: includes trailing slash (/)
+    log.info("chosen apache mirror is " + apache_mirror_url_front)
+
+    for (client, _) in config.items():
+        (remote,) = ctx.cluster.only(client).remotes.keys()
+        test_dir=teuthology.get_testdir(ctx)
+        current_version = get_kafka_version(config)
+
+        kafka_file =  kafka_prefix + current_version + '.tgz'
+
+        link1 = '{apache_mirror_url_front}/kafka/'.format(apache_mirror_url_front=apache_mirror_url_front) + \
+            current_version + '/' + kafka_file
+        ctx.cluster.only(client).run(
+            args=['cd', '{tdir}'.format(tdir=test_dir), run.Raw('&&'), 'wget', link1],
+        )
+
+        ctx.cluster.only(client).run(
+            args=['cd', '{tdir}'.format(tdir=test_dir), run.Raw('&&'), 'tar', '-xvzf', kafka_file],
+        )
+
+        kafka_dir = get_kafka_dir(ctx, config)
+        # create config for second broker
+        second_broker_config_name = "server2.properties"
+        second_broker_data = "{tdir}/data/broker02".format(tdir=kafka_dir)
+        second_broker_data_logs_escaped = "{}/logs".format(second_broker_data).replace("/", "\/")
+
+        ctx.cluster.only(client).run(
+            args=['cd', '{tdir}'.format(tdir=kafka_dir), run.Raw('&&'), 
+             'cp', '{tdir}/config/server.properties'.format(tdir=kafka_dir), '{tdir}/config/{second_broker_config_name}'.format(tdir=kafka_dir, second_broker_config_name=second_broker_config_name), run.Raw('&&'), 
+             'mkdir', '-p', '{tdir}/data'.format(tdir=kafka_dir)
+            ],
+        )
+
+        # edit config
+        ctx.cluster.only(client).run(
```

Comment:
this is failing with:
```
2024-12-18T09:29:29.716 DEBUG:teuthology.orchestra.run.smithi027:> sed '-i""' '"s/broker.id=0/broker.id=1/g"' /home/ubuntu/cephtest/kafka_2.13-3.8.1/config/server2.properties && sed '-i""' '"s/listeners=PLAINTEXT:\/\/:19092/listeners=PLAINTEXT:\/\/localhost:19092 /g"' /home/ubuntu/cephtest/kafka_2.13-3.8.1/config/server2.properties && sed '-i""' '"s/#advertised.listeners=PLAINTEXT:\/\/:9092/advertised.listeners=PLAINTEXT://localhost:19092/g"' /home/ubuntu/cephtest/kafka_2.13-3.8.1/config/server2.properties && sed '-i""' '"s/log.dirs=\/tmp\/kafka-logs/log.dirs=\/home\/ubuntu\/cephtest\/kafka_2.13-3.8.1\/data\/broker02\/logs"' /home/ubuntu/cephtest/kafka_2.13-3.8.1/config/server2.properties
2024-12-18T09:29:29.761 DEBUG:teuthology.orchestra.run:got remote process result: 1
2024-12-18T09:29:29.762 INFO:teuthology.orchestra.run.smithi027.stderr:sed: -e expression #1, char 1: unknown command: `"'
2024-12-18T09:29:29.762 ERROR:teuthology.contextutil:Saw exception from nested tasks
```

Reply:
see: https://pulpito.ceph.com/yuvalif-2024-12-18_09:04:51-rgw:notifications-wip-yuval-kafka_multiple_brokers-distro-default-smithi/

Reply:
I removed empty space after -i params

Reply:
still failing: https://pulpito.ceph.com/yuvalif-2024-12-18_11:28:43-rgw:notifications-wip-yuval-kafka_multiple_brokers-distro-default-smithi/